driver-core/src/main/com/mongodb/internal/connection/DefaultServerMonitor.java

```diff
@@ -165,8 +192,7 @@ public void run() {
                     logStateChange(previousServerDescription, currentServerDescription);
                     sdamProvider.get().update(currentServerDescription);
 
-                    if (((connection == null || shouldStreamResponses(currentServerDescription))
```

Comment:
is this check "connection == null" not needed anymore?

Reply:
I thought I had proven to myself that the check was redundant by transforming to

```
                    boolean shouldStream = currentServerDescription.getTopologyVersion() != null;
                    if (((connection == null || shouldStream)
                            && shouldStream && currentServerDescription.getType() != UNKNOWN)
                            || (connection != null && connection.hasMoreToCome())
                            || (currentServerDescription.getException() instanceof MongoSocketException
                            && previousServerDescription.getType() != UNKNOWN)) {
                        continue;
                    }
```

and then allowing Intellij to simplify the expression, but when I tried it again IntelliJ didn't offer the simplification.

This is a truly horrible conditional, and has been the source of at least one bug that I remember.

Reply:
> is this check "connection == null" not needed anymore?

It had no effect, so I removed it. The following demonstrates why the `connection == null` expression had no effect.

Here is the original conditional, but reformatted to simplify perception:

```java
(
    (connection == null || shouldStreamResponses(currentServerDescription))
    && currentServerDescription.getTopologyVersion() != null
    && currentServerDescription.getType() != UNKNOWN
)
||
(connection != null && connection.hasMoreToCome())
||
(
    currentServerDescription.getException() instanceof MongoSocketException
    && previousServerDescription.getType() != UNKNOWN
)
```

It consists of three Boolean expressions ORed (`||`) together. Only the first one of those is modified in the PR:

1. Let's alias the `connection == null` expression as `x`.
2. Note that the expressions `shouldStreamResponses(currentServerDescription)` and `currentServerDescription.getTopologyVersion() != null` are equivalent. Let's alias them as `s`.

Now the first ORed Boolean expression can be written as

```java
(
    (x || s)
    && s
    && currentServerDescription.getType() != UNKNOWN
)
```

If `s` is `false`, this expression evaluates to `false`. If `s` is `true`, this expression evaluates to the same value as `currentServerDescription.getType() != UNKNOWN`. As you can see, `x` affects the outcome is neither case, and the expression can be simplified to

```java
(
    s
    && currentServerDescription.getType() != UNKNOWN
)
```

Reply:
Thanks for the clarification, @stIncMale. It is indeed redundant check.

i am curious about the original purpose of this check, given it wasn't functional. My assumption was that it aimed to bypass `waitForNext()` on lookup failure as connection is nullified in lookup method. However, the presence of ` (currentServerDescription.getException() instanceof MongoSocketException
                            && previousServerDescription.getType() != UNKNOWN)`seems to cover scenario aligning with the spec, which mandates immediate retry on network errors when the server was previously in a known state, without unnecessary delays:

#. If this was a network error and the server was in a known state before the error, the client MUST NOT sleep and MUST begin the next check immediately.

Resolving the thread.